File: csharp/scim-UserList/scim-UserList/scim-UserList.cs
using Newtonsoft.Json.Linq;
using RestSharp;
using System;
using System.Linq;

namespace IonicAPISample
{
    class SCIMUserList
    {



        public class IonicQueryParameters
        {
            public IonicQueryParameters()
            {
                or = null;
                domainUpn = null;
                email = null;
                enabled = null;
                externalId = null;
                groups = null;
                group_empty = null;
                roles = null;
                createdTs = null;
                updatedTs = null;
                attributes = null;

                startIndex = null;
                count = null;
                skip = null;
                limit = null;
            }

            // Query filter parameters
            public bool? or { get; set; }

            public string domainUpn { get; set; }
            public string email { get; set; }
            public bool? enabled { get; set; }
            public string externalId { get; set; }
            public string groups { get; set; }
            public bool? group_empty { get; set; }
            public string roles { get; set; }
            public int? createdTs { get; set; }
            public int? updatedTs { get; set; }

            public string attributes { get; set; }

            // Pagination parameters
            public int? startIndex { get; set; }
            public int? count { get; set; }
            public int? skip { get; set; }
            public int? limit { get; set; }
        }

        static void Main(string[] args)
        {
            // Load all needed info from user's config file
            //
            if (!SampleConfig.LoadConfig())
            {
                Console.WriteLine($"Error loading config from:  {SampleConfig.ConfigFile}");
                Console.WriteLine();

                // Keep console app open to see results
                Console.WriteLine("\nPress return to exit.");
                Console.ReadKey();
            }

            // Output header
            Console.WriteLine($"User List");
            Console.WriteLine($"  Host:    {SampleConfig.APIURL}");
            Console.WriteLine($"  Tenant:  {SampleConfig.TenantID}");
            Console.WriteLine();

            //
            // REST:  SCIM/List Users
            //
            var client = new RestClient($"{SampleConfig.APIURL}/v2/{SampleConfig.TenantID}/scim/Users");

            var reqParms = new IonicQueryParameters();
            reqParms.attributes = "name,emails";
            reqParms.startIndex = 1;
            reqParms.count = 50;        // page size

            // "Paginated" request (50/per request)
            int totalUsers = 2;         // start with value to get inside loop - get actual value from response field
            string lastUserId = "";     // place to save a user ID
            while (reqParms.startIndex < totalUsers)
            {
                // Setup new request
                var request = new RestRequest(Method.GET);
                request.AddHeader("Authorization", SampleConfig.AuthHeader);
                request.AddObject(reqParms);

                // REST call + handle failure
                //
                IRestResponse response = client.Execute(request);
                if (!response.IsSuccessful)
                {
                    Console.WriteLine($"List Users ERROR:  Response = {response.Content}");
                    Console.WriteLine();

                    // Keep console app open to see results
                    Console.WriteLine("\nPress return to exit.");
                    Console.ReadKey();
                    return;
                }

                // Fetch "page" of user list
                JObject json = JObject.Parse(response.Content);

                totalUsers = (int)json["totalResults"];     // real #
                JArray userList = (JArray)json["Resources"];

                // Write user info to console
                int userIdx = (int)json["startIndex"];
                foreach (JObject userObj in userList)
                {
                    // Save last user's id
                    lastUserId = (string)userObj["id"];

                    JArray emailList = (JArray)userObj["emails"];
                    if (emailList != null && emailList.Count > 0)
                    {
                        Console.WriteLine($"[{userIdx}]:  {userObj["id"]} - {userObj["name"]["formatted"]}  {emailList[0]["value"]}");
                    }
                    else
                    { 
                        Console.WriteLine($"[{userIdx}]:  {userObj["id"]} - {userObj["name"]["formatted"]}");
                    }
                    userIdx++;
                }

                reqParms.startIndex += reqParms.count;
            }

            //
            // REST:  SCIM/Fetch User
            //
            client = new RestClient($"{SampleConfig.APIURL}/v2/{SampleConfig.TenantID}/scim/Users/{lastUserId}");

            // Using parameters class from before, but only "attributes" works for Fetch User request
            reqParms = new IonicQueryParameters();

            var request2 = new RestRequest(Method.GET);
            request2.AddHeader("Authorization", SampleConfig.AuthHeader);
            request2.AddObject(reqParms);

            // REST call + handle failure
            //
            IRestResponse response2 = client.Execute(request2);
            if (!response2.IsSuccessful)
            {
                Console.WriteLine("Fetch User ERROR:");
                Console.WriteLine($"  Response = {response2.Content}");
                Console.WriteLine();

                // Keep console app open to see results
                Console.WriteLine("\nPress return to exit.");
                Console.ReadKey();
                return;
            }

            JObject userJson = JObject.Parse(response2.Content);

            // Output some interesting stuff from user
            Console.WriteLine();
            Console.WriteLine($"First name: {userJson["name"]["givenName"]}");
            Console.WriteLine($"Last name:  {userJson["name"]["familyName"]}");
            Console.WriteLine($"Created:    {userJson["meta"]["created"]}");

            JArray roleList = (JArray)userJson["roles"];
            if (roleList != null)
            {
                Console.WriteLine("Roles:");
                foreach (JObject role in roleList)
                    Console.WriteLine($"   {role["value"]}");
            }

            // Keep console app open to see results when run from Visual Studio
            Console.WriteLine();
            Console.WriteLine("\nPress return to exit.");
            Console.ReadKey();
        }
    }
}

File: java/scimUserList/src/main/java/com/ionic/samples/api/ScimUserList.java
/*
 * (c) 2019 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

package com.ionic.samples.api;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import org.json.JSONArray;
import org.json.JSONObject;


final class ScimUserList {

    private ScimUserList() { }

    public static void main(final String[] args) {
        // Load all needed info from user's config file
        //
        if (!SampleConfig.loadConfig()) {
            System.out.println("Error loading config from:  " + SampleConfig.getConfigFile());
            System.out.println();

            return;
        }

        // Output header
        System.out.println("User List");
        System.out.println("  Host:    " + SampleConfig.getApiUrl());
        System.out.println("  Tenant:  " + SampleConfig.getTenantID());
        System.out.println();

        // client object for REST calls
        final OkHttpClient client = new OkHttpClient();

        // Structure with properties for REST API calls
        //      - initial setup for List Users call
        //
        final Map<String, String> properties = new HashMap();
        properties.put("attributes", "name,emails");
        properties.put("startIndex", "1");

        final int count = 50;
        properties.put("count", String.valueOf(count));

        // Loop to get user list in "pages"
        int totalUsers = 2;
        String lastUserId = "";
        while (Integer.parseInt(properties.get("startIndex")) < totalUsers) {
            // API call #1:  List Users
            //
            final HttpUrl.Builder urlBuilder = HttpUrl.parse(SampleConfig.getApiUrl() + "/v2/"
                    + SampleConfig.getTenantID() + "/scim/Users").newBuilder();
            setUrlProperties(urlBuilder, properties);
            final Request request = new Request.Builder()
                    .url(urlBuilder.build().toString())
                    .addHeader("Authorization", SampleConfig.getAuthHeader())
                    .build();

            try (Response response = client.newCall(request).execute()) {
                if (!response.isSuccessful()) {
                    System.out.println("User List error.  Response:  " + response.body().string());
                    System.out.println();

                    return;
                }

                // Parse response JSON
                final JSONObject json = new JSONObject(response.body().string());

                totalUsers = (int) json.get("totalResults");

                final JSONArray userList = (JSONArray) json.get("Resources");
                int userIndex = (int) json.get("startIndex");
                for (final Object userObj : userList) {
                    final JSONObject user = (JSONObject) userObj;

                    // Save last user's ID
                    lastUserId = (String) user.get("id");

                    final JSONArray emailList = (JSONArray) user.get("emails");
                    if (emailList != null && emailList.length() > 0) {
                        System.out.println("[" + userIndex + "]:  " + user.get("id").toString() + " - "
                                + ((JSONObject) user.get("name")).get("formatted").toString() + " "
                                + ((JSONObject) emailList.get(0)).get("value").toString());
                    } else {
                        System.out.println("[" + userIndex + "]:  " + user.get("id").toString() + " - "
                                + ((JSONObject) user.get("name")).get("formatted").toString());
                    }

                    userIndex++;
                }
            } catch (IOException ex) {
                System.out.println("User list request failure: " + ex.getMessage());
                break;
            }

            // Advance start index by amount of count
            final int startIndex = Integer.parseInt(properties.get("startIndex")) + count;
            properties.put("startIndex", String.valueOf(startIndex));
        }

        // API call #2:  Fetch User
        //
        final HttpUrl.Builder urlBuilder = HttpUrl.parse(SampleConfig.getApiUrl() + "/v2/" + SampleConfig.getTenantID()
                + "/scim/Users/" + lastUserId).newBuilder();
        final Request request = new Request.Builder()
                .url(urlBuilder.build().toString())
                .addHeader("Authorization", SampleConfig.getAuthHeader())
                .build();

        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                System.out.println("Fetch User error.  Response:  " + response.body().string());
                System.out.println();

                return;
            }

            // Parse response JSON
            final JSONObject json = new JSONObject(response.body().string());

            // Output some interesting stuff from user
            System.out.println();
            final JSONObject name = (JSONObject) json.get("name");

            System.out.println("First name: " + name.get("givenName"));
            System.out.println("Last name:  " + name.get("familyName"));
            System.out.println("Created:    " + ((JSONObject) json.get("meta")).get("created"));

            final JSONArray roleList = (JSONArray) json.get("roles");
            if (roleList != null) {
                System.out.println("Roles:");
                for (final Object roleObj : roleList) {
                    final JSONObject role = (JSONObject) roleObj;

                    System.out.println("   " + role.get("value"));
                }
            }
        } catch (Exception ex) {
            System.out.println("Unique users request failure: " + ex.getMessage());
        }
    }

    /**
     * (Re-)Setup URL builder with collection of properties.
     * @param builder HTTP URL builder object
     * @param properties Map of properties for URL
     */
    public static void setUrlProperties(final HttpUrl.Builder builder, final Map<String, String> properties) {
        for (final Map.Entry<String, String> prop : properties.entrySet()) {
            builder.removeAllQueryParameters(prop.getKey());
            builder.addQueryParameter(prop.getKey(), prop.getValue());
        }
    }
}
